### ISSUE TITLE
__iter__ type checks not working as expected?

### ISSUE THREAD

**Issue by Dedoig** [JukkaL/mypy#660]
```
def f():
	return 9

for i in f():
	for j in f:
		print(i, j)
```

Shouldn't both for-loop lines be type errors due to __iter__? Mypy only reports the second one.

**Comment by spkersten** [comment]
When you don't specify a return type for a function, mypy does't do any type checking on it and its return type is inferred to be the `Any` type, which is compatible with every other type. Therefore, `for i in f(): ...` doesn't result in type errors. (`for j in f` does raise an error because `f` is still a function, which do not support iteration over.)

**Comment by spkersten** [comment]
If you change your function to:
```python
def f() -> int:
    return 9
```
you do get a type error for both loops.

**Comment by JukkaL** [comment]
Yeah, this is the expected behavior.

However, maybe we should have a strict type checking mode that complains about missing type annotations...
